### ISSUE TITLE
Does `within_cluster_degenes` work?

### ISSUE THREAD

**Issue by vals** [YosefLab/scVI#561]
Hi,

I hadn't seen this function documented before and was working on a similar functionality until I realized it had already been implemented!

When I have been testing it I have gotten some odd results however. I get the same DE results for every cluster between the two conditions. It is almost as if the `cell_labels` mask is ignored?

Just to be sure I understand what it does correctly. I should be able to take the data from https://doi.org/10.1038/s41593-017-0029-5, integrate/harmonize the 23 mice, then be able to detect e.g. that the `ExcL23` cluster has Nr4a2 increased between the 0h and 1h condition? But the same not being true in the `Olig_1` cluster?

Thanks!
/Valentine

**Comment by adamgayoso** [comment]
I actually just fixed this and merged it into master a few days ago. The issue was that the states were converted to ints when they should have been bool so they were basically only comparing cell 0 to cell 1. For now, please make sure cell labels is array of `int` and states is array of `bool`.

**Comment by vals** [comment]
Hi,

I had noticed and downloaded the master branch and installed. But still had issues. I think this was because I had saved a fitted model that I loaded `torch.load()`.

Today I fitted a new model, saved that one, and loaded it to do the DE, and now I'm getting more reasonable results.

I'm not certain how the models get saved when using `torch.save()`. If I make an object and save it, does the saved file include the code it uses in its methods? Or does it have references to the location of the code that defines the method?

In other words, if I save a model, will it only be able to use the code that was defined while I saved it?

**Comment by vals** [comment]
By the way, the functionality of `within_cluster_degenes` seem to work pretty well with this test! Here is a screenshot from the Hrvatin paper I linked above, showing which genes they find are activated in which cell types:
![image](https://user-images.githubusercontent.com/668803/73214293-dfa2c700-411f-11ea-92c2-65d56ad06828.png)

And here are the results I got using `within_cluster_degenes` with `mode='change'`:
![image](https://user-images.githubusercontent.com/668803/73214447-37413280-4120-11ea-9d54-a0681d172c22.png)
The red genes are ones in the original heatmap.

**Comment by adamgayoso** [comment]
Looks cool! 

We recommend saving and loading with the first method described here:

https://pytorch.org/tutorials/beginner/saving_loading_models.html

Since it's more resilient to refactors. 

Though it's unclear to me why you needed to retrain, unless you saved and loaded the posterior object?

**Comment by vals** [comment]
Yes I saved the `Posterior` object, I saw you save the `Trainer` object? I didn't want to need to also save the `GeneDataset` object. Would it had worked out if I had saved the `Trainer` object? Can I easily make a posterior object without rebuilding the `GeneDataset` object that I mass to the `.create_posterior()` method?

**Comment by adamgayoso** [comment]
We save the model only typically and then rebuild the gene dataset (a pain, I know...)

From the basic tutorial

```python
torch.save(trainer.model.state_dict(), '%s/vae.pt' % save_path)
```
trainer.model and the model point to the same object.

I suppose saving the posterior has the greatest ease of use, but since the methods are bound on reload (I think this is the right term?), you'd run into the issue you had where the posterior you loaded had an older version of the method.

I guess another alternative way to do things is to pickle the gene expression dataset (scVI object) and the model (model state dict way) and then use both of those to build your own posterior (using it's own init method, not the create posterior method in the trainer).

**Comment by adamgayoso** [comment]
Also, we could just write a posterior save and load method... that would be a bit smarter than just pickling the whole object.

**Comment by vals** [comment]
This sounds like a great idea!

In my current workflow I launch a new AWS EC2 instance when I need to do an analysis, and it was getting a bit silly to make and train a model every time I wanted to do an analysis. An easy way to save and load both dataset and model would make it much easier to perform off-the-cuff analysis.

**Comment by adamgayoso** [comment]
Yeah for our current project I'm reassembling the dataset in every notebook. I'll create an issue for this.

On another note, @vals how are calculating the LFC in those volcano plots? It's likely not clear in our documentation, but the column called "mean" in the `df` returned by the DE functions is the mean LFC (here LFC is a random variable). I'll note that we also return credible intervals for the LFC in the dataframe .

**Comment by vals** [comment]
Oh I didn't realize that. I calculated it using the `scale1` and `scale2` column using log2 differences like in the default change function. I was plotting the credible intervals as well for a while but it got pretty crowded. I'll have a look what it looks like with the mean column.

**Comment by adamgayoso** [comment]
Great -- going back to the data loading a bit I'd highly recommend running `model.eval()` every time you load. The trainer puts the model in eval after training, but it's not clear with the way you save the posterior if it will stay in eval mode. We had some weird results by not putting it in eval mode accidentally one day so it's definitely something to keep in mind.

**Comment by adamgayoso** [comment]
> Oh I didn't realize that. I calculated it using the scale1 and scale2 column using log2 differences like in the default change function. I was plotting the credible intervals as well for a while but it got pretty crowded. I'll have a look what it looks like with the mean column.

So yeah this would be like computing the LFC of the expectation instead of the expectation of the LFC.

**Comment by vals** [comment]
I remade the plot using the 'mean' column, it looks like a large number of outliers went away (though I guess this could also be sampling noise). Compare e.g. `Int_Sst_1`:
![image](https://user-images.githubusercontent.com/668803/73276590-efb9b580-41b6-11ea-9c29-14029ea8d7cb.png)

**Comment by adamgayoso** [comment]
Looks great! Thanks for following up. I'm going to close the issue for now, but if you have any recommendations for making DE more user-friendly please feel free to open another issue, @vals
